### PR TITLE
Improves the haskell solution by changing to a set

### PR DESCRIPTION
**Pull request by jchildren**
Hi, I'm one of Johan's colleagues who pointed me at your repository.

Been messing around with trying to speed up the Haskell implementation for a couple of days and this is the best I've managed so far. It can solve the problem in less than 5 seconds on my older processor (haven't tested others for comparison yet) and doesn't use any external libraries (other than `containers` which is part of the haskell project) or use `Data.Vector`. It may been sort of cheating because it doesn't implement the exact same algorithm, but it is still a brute force solution. It would also be interesting to see how it performs with the `-fllvm` option enabled, but I doubt it will make much difference (I suspect there are other compile flags which will get arcane performance boosts out of ghc).

I don't know if you want merge requests or if you want to just copy the code in, I'm easy either way.

I also have a fortran solution in the works, but my quicksort implementation keeps segfaulting :/

### CODE REVIEW

**Comment by clux** [comment]
Awesome, thanks a lot. :open_mouth:

It looks great. I don't think you are changing the algorithm sufficiently for it to be considered cheating, it would only be the (internally done) uniqueness from before the invalid filter, but that's probably doesn't make it easier for itself. Besides, `go` uses the same trick for lack of better ways. Sets are certainly used elsewhere.

It's certainly a lot cleaner than what I tried to do with `Data.Vector` which needed me to mix mutables and immutables in pretty expansive ways.

Anyway, I'll merge and benchmark and try the flag you suggested. Thanks again.

I'd be happy to take and bench a fortran solution provided you can get it to print each stage as verification (see [test.sh](./test.sh)) because I don't really know how to code review fortran unfortunately :sweat:

File: forest.hs [hunk 1]
```diff
@@ -1,40 +1,41 @@
...
-import Data.List
+module Main where
+
+import qualified Data.Set as S
 import System.Environment
 
 data Forest = Forest
   { goats :: !Int
   , wolves :: !Int
   , lions :: !Int
```
Comment (clux) [inline forest.hs]: oh, quick question; you've removed bang patterns, does these bangs still mean anything without the extension?

Comment (jchildren) [inline forest.hs]: Those are strictness annotations for the records, they basically make all pattern matching use bang patterns without needing to write them out? That's as far as I understand at least. See http://book.realworldhaskell.org/read/profiling-and-optimization.html about halfway down the page titled "Bang Patterns" then the section below

**Comment by jchildren** [comment]
>I'd be happy to take and bench a fortran solution provided you can get it to print each stage as verification (see test.sh) because I don't really know how to code review fortran unfortunately

You'd probably be surprised, it's fortran 2003 so it looks a bit like ruby.

File: forest.hs (hunk 1 above)
Comment (clux) [inline forest.hs]: Ah, I see. I tried taking it out and performance dropped by a almost factor of two, so definitely worth keeping :)

**Comment by clux** [comment]
Oh, okay, well I'm up for a challenge.

At any rate, your Haskell one now runs in 3s locally. Half the number of lines of go, double the speed. This is more numbers I expected.

**Comment by clux** [comment]
I also tried `ghc -fllvm -O3 forest.rs`, but it didn't seem to change the run speed at all.

**Comment by jchildren** [comment]
Fair enough, I think `-fllvm` was actually only going to help with `Data.Vector`, but the cost of converting between sets and vectors was too painful. If [this issue](https://github.com/haskell/vector/issues/145) gets resolved and a fast `nub` is implemented for vectors without conversions a sorting based approach might be more efficient again.

File: forest.hs [hunk 2]
```diff
@@ -1,40 +1,41 @@
...
-solve x = solve' [x]
+solve :: Forest -> S.Set Forest
+solve = solve' . S.singleton
   where
   solve' xs | cond xs = solve' $ mutate xs
-            | otherwise = filter is_stable xs
-  cond [] = False
-  cond xs = null . filter is_stable $ xs
+            | otherwise = S.filter isStable xs
+  cond vec | S.null vec = False
```
Comment (jchildren) [inline forest.hs]: whoops this should be `cond set` now not `cond vec`

Comment (clux) [inline forest.hs]: ah, np, i'll change that now

**Pull request merged**